File: web/src/types/certificacion.ts
import { z } from 'zod';

export const ESTADOS_CERTIFICACION = ['vigente', 'anulada', 'ejecutada'] as const;
export type EstadoCertificacion = (typeof ESTADOS_CERTIFICACION)[number];

export const certificacionSchema = z.object({
  entidad_id: z.string().uuid('Seleccione una entidad'),
  disponibilidad_id: z.string().uuid('Seleccione una disponibilidad aprobada'),
  numero_certificado: z.string().min(3, 'Mínimo 3 caracteres').max(50),
  monto_certificado: z.coerce.number().min(0.01, 'Debe ser mayor a 0'),
  fecha_certificacion: z.string().min(1, 'La fecha es requerida'),
  anio_fiscal: z.coerce.number().min(2000).max(2100),
  observaciones: z.string().optional().or(z.literal('')),
});

export type CertificacionFormValues = z.infer<typeof certificacionSchema>;

export interface CertificacionEntity {
  id: string;
  entidad_id: string;
  disponibilidad_id: string;
  numero_certificado: string;
  monto_certificado: number;
  fecha_certificacion: string;
  anio_fiscal: number;
  estado: EstadoCertificacion;
  observaciones?: string;
  certificado_por?: string;
  creado_en: string;
  actualizado_en?: string;
  // JOINs
  entidad_nombre?: string;
  disponibilidad_numero?: string;
  disponibilidad_monto_aprobado?: number;
  certificador_nombre?: string;
}

export interface CertificacionPaginatedResponse {
  data: CertificacionEntity[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export interface CertificacionListParams {
  page: number;
  limit: number;
  search?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  entidad_id?: string;
  estado?: string;
  anio_fiscal?: number;
}

File: api/app/schemas/certificacion.py
from pydantic import BaseModel, ConfigDict, field_validator
from uuid import UUID
from datetime import date, datetime
from decimal import Decimal

ESTADOS_CERTIFICACION = ["vigente", "anulada", "ejecutada"]


class CertificacionBase(BaseModel):
    entidad_id: UUID
    disponibilidad_id: UUID
    numero_certificado: str
    monto_certificado: Decimal
    fecha_certificacion: date
    anio_fiscal: int
    observaciones: str | None = None
    estado: str = "vigente"

    @field_validator("numero_certificado")
    @classmethod
    def validar_numero(cls, v: str) -> str:
        v = v.strip()
        if len(v) < 3:
            raise ValueError("El número de certificado es muy corto")
        return v

    @field_validator("monto_certificado")
    @classmethod
    def validar_monto(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("El monto debe ser mayor a 0")
        return v

    @field_validator("estado")
    @classmethod
    def validar_estado(cls, v: str) -> str:
        if v not in ESTADOS_CERTIFICACION:
            raise ValueError(f"Estado debe ser uno de: {', '.join(ESTADOS_CERTIFICACION)}")
        return v


class CertificacionCreate(CertificacionBase):
    pass


class CertificacionUpdate(BaseModel):
    observaciones: str | None = None
    estado: str | None = None


class CertificacionOut(CertificacionBase):
    id: UUID
    certificado_por: UUID | None = None
    creado_en: datetime
    actualizado_en: datetime | None = None
    
    # Campos relacionados (JOINs)
    entidad_nombre: str | None = None
    disponibilidad_numero: str | None = None
    disponibilidad_monto_aprobado: Decimal | None = None
    certificador_nombre: str | None = None
    
    model_config = ConfigDict(from_attributes=True)